browser/net/brave_localhost_permission_network_delegate_helper.cc

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2023 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/browser/net/brave_localhost_permission_network_delegate_helper.h"
+
+#include <array>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "brave/browser/brave_browser_process.h"
+#include "chrome/browser/content_settings/host_content_settings_map_factory.h"
+#include "chrome/browser/profiles/profile.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "components/prefs/pref_service.h"
+#include "content/public/browser/browser_context.h"
+#include "content/public/browser/permission_controller_delegate.h"
+#include "content/public/browser/web_contents.h"
+#include "net/base/net_errors.h"
+#include "net/base/url_util.h"
+#include "third_party/blink/public/common/permissions/permission_utils.h"
+
+namespace brave {
+
+void OnPermissionRequestStatus(
+    content::WebContents* contents,
+    const std::vector<blink::mojom::PermissionStatus>& permission_statuses) {
+  DCHECK_EQ(1u, permission_statuses.size());
+  if (contents &&
+      permission_statuses[0] == blink::mojom::PermissionStatus::GRANTED) {
+    contents->GetController().Reload(content::ReloadType::NORMAL, true);
+  }
+}
+
+bool IsLocalhostRequest(const GURL& request_url,
+                        const GURL& request_initiator_url) {
+  return request_initiator_url.is_valid() && request_url.is_valid() &&
+         net::IsLocalhost(request_url) &&
+         !net::IsLocalhost(request_initiator_url);
+}
```

Comment:
There was a lot of discussion about this with @bridiver and others. We've decided to do this in a follow up because it would be complicated to handle correctly.

Reply:
Are we sure the proper support for the proper localhost permission won't look like a full rewrite?

This is a nice feature, but having it in the current state is misleading IMO. I think if we're merging it as is, we need to be explicit about this in the changelog.

Reply:
Great question, one that I was hesitant about as well initially. But I don't think it will be a rewrite and instead will be incremental. 

Currently, this feature essentially throttles subresource requests for resources hosted on the user's localhost purely based on the request URL, before any connection is made (and then permissions them etc). For a complete solution, we also want the IP address of the target endpoint to check if it's localhost. But even once we have that, we would still want to throttle the localhost requests we can before the initial connection is made, transport is established etc. 

The feature will be off by default in any case.

Reply:
I can update the issue description to be explicit about this.

Reply:
Also @pes10k pointed out that we can already use filter lists to block requests to known bad domains.